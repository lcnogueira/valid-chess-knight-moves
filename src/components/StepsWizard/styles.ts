import styled, { css } from 'styled-components'

export const Container = styled.div`
  max-width: 70rem;
  width: 100%;
`

export const ButtonsContainer = styled.div`
  ${({ theme: { screens } }) => css`
    display: flex;
    flex-direction: column;
    align-items: stretch;
    justify-content: stretch;
    margin-bottom: 2rem;

    @media (min-width: ${screens.small}) {
      flex-direction: row;
      justify-content: center;
    }
  `}
`

export const NavigationButton = styled.button`
  ${({ theme: { colors, border } }) => css`
    background-color: ${colors.secondary};
    color: ${colors.white};
    margin: 1rem;
    min-width: 15rem;
    padding: 1rem 2rem;
    border-radius: ${border.radius};
    font-size: 1.6rem;

    :hover {
      background-color: ${colors.white};
      color: ${colors.secondary};
    }
  `}
`

export const Steps = styled.span`
  margin: 1rem;
  font-size: 2rem;
`
