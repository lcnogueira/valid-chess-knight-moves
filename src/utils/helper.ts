export const lines = ['8', '7', '6', '5', '4', '3', '2', '1']
export const columns = ['A', 'B', 'C', 'D', 'E', 'F', 'G', 'H']

export const isValidNumber = (number: string) => lines.includes(number)

export const isValidLetter = (letter: string) =>
  columns.includes(letter.toUpperCase())

/**
 * The position is valid if componse by letter followed by one number and they both belong to the chess board limit.
 * @param position The position the user clicked on
 */
export const isAlgebraicNotation = (position: string) => {
  if (
    /^\D\d$/.test(position) &&
    isValidLetter(position.charAt(0)) &&
    isValidNumber(position.charAt(1))
  ) {
    return true
  }
  return false
}

/**
 * Returns the knight possible positions.
 * @param position The position the user clicked on
 */
export const findKnightPossibleMoves = (position: string): string[] => {
  if (position === 'A8') {
    return ['B6', 'C7']
  } else {
    return []
  }
}
