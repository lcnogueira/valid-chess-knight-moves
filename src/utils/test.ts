import {
  isAlgebraicNotation,
  isValidPosition,
  isValidIndex,
  findKnightPossibleMoves
} from './helper'

describe('helper methods', () => {
  test(`isAlgebraicNotation returns true if it's a valid notation`, () => {
    expect(isAlgebraicNotation('A1')).toBeTruthy()
    expect(isAlgebraicNotation('a1')).toBeTruthy()
    expect(isAlgebraicNotation('I1')).toBeTruthy()
    expect(isAlgebraicNotation('1A')).toBeFalsy()
    expect(isAlgebraicNotation('A11')).toBeFalsy()
  })

  test(`isValidPosition returns true if it's a valid notation and is in the chess board limit`, () => {
    expect(isValidPosition('A1')).toBeTruthy()
    expect(isValidPosition('A8')).toBeTruthy()
    expect(isValidPosition('h1')).toBeTruthy()
    expect(isValidPosition('H8')).toBeTruthy()
    expect(isValidPosition('A9')).toBeFalsy()
    expect(isValidPosition('H9')).toBeFalsy()
  })

  test('isValidIndex returns true if the index is equal or less than the array size', () => {
    expect(isValidIndex(0)).toBeTruthy()
    expect(isValidIndex(7)).toBeTruthy()
    expect(isValidIndex(-1)).toBeFalsy()
    expect(isValidIndex(8)).toBeFalsy()
  })

  test('findKnightPossibleMoves returns the posible positions where the knight can move', () => {
    expect(findKnightPossibleMoves('A8').sort()).toEqual(
      [
        'A4',
        'A6',
        'A8',
        'B5',
        'B6',
        'C4',
        'C7',
        'C8',
        'D5',
        'D7',
        'E6',
        'E8'
      ].sort()
    )
  })
})
