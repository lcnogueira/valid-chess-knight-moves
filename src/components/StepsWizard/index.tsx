import { useState } from 'react'

import * as S from './styles'

interface Props {
  steps: Array<React.FC>
  showSteps?: boolean
}

const StepsWizard: React.FC<Props> = ({ steps, showSteps = true }) => {
  const [currentStep, setCurrentStep] = useState(0)

  const onPrev = () => setCurrentStep((prev) => prev - 1)

  const onNext = () => setCurrentStep((prev) => prev + 1)

  return (
    <S.Container>
      <S.ButtonsContainer>
        <S.NavigationButton onClick={onPrev} disabled={currentStep === 0}>
          Previous
        </S.NavigationButton>
        <S.NavigationButton
          onClick={onNext}
          disabled={currentStep === steps.length - 1}
        >
          Next
        </S.NavigationButton>
      </S.ButtonsContainer>
      {showSteps && (
        <S.Steps>{`Step ${currentStep + 1}/${steps.length}`}</S.Steps>
      )}
      {steps[currentStep]}
    </S.Container>
  )
}

export default StepsWizard
