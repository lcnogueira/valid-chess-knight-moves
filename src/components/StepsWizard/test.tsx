import { render, screen, userEvent } from 'tests/utils'

import StepsWizard from '.'

const step1 = () => <div data-testid="step">Step 1</div>
const step2 = () => <div data-testid="step">Step 2</div>

const DefaultStepsWizard = ({ showSteps = false }) => (
  <StepsWizard steps={[step1, step2]} showSteps={showSteps} />
)

describe('<StepsWizard />', () => {
  it('renders correctly with props', () => {
    const { container } = render(<DefaultStepsWizard />)

    expect(container).toMatchSnapshot()
  })

  it('changes the current step when clicking the buttons', () => {
    render(<DefaultStepsWizard />)

    const next = screen.getByRole('button', { name: /next/i })
    const previous = screen.getByRole('button', { name: /previous/i })

    expect(next).toBeEnabled()
    expect(previous).toBeDisabled()
    expect(screen.getAllByTestId('step')).toHaveLength(1)
    expect(screen.getByTestId('step')).toHaveTextContent('Step 1')

    userEvent.click(next)

    expect(next).toBeDisabled()
    expect(previous).toBeEnabled()
    expect(screen.getAllByTestId('step')).toHaveLength(1)
    expect(screen.getByTestId('step')).toHaveTextContent('Step 2')

    userEvent.click(previous)

    expect(next).toBeEnabled()
    expect(previous).toBeDisabled()
    expect(screen.getAllByTestId('step')).toHaveLength(1)
    expect(screen.getByTestId('step')).toHaveTextContent('Step 1')
  })

  it('shows the steps in case showSteps is true', () => {
    render(<DefaultStepsWizard showSteps={true} />)

    expect(screen.getByText(/step 1 \/ 2/i)).toBeInTheDocument()

    userEvent.click(screen.getByRole('button', { name: /next/i }))

    expect(screen.getByText(/step 2 \/ 2/i)).toBeInTheDocument()
  })
})
