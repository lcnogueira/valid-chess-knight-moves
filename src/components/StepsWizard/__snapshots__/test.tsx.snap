// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<StepsWizard /> renders correctly with props 1`] = `
.c2 {
  margin: 1rem;
  min-width: 15rem;
  padding: 1.5rem 2.5rem;
  border-radius: 0.4rem;
  font-size: 1.6rem;
  font-weight: bold;
  border: yellow;
  background-color: #ffa500;
  color: #fff;
  opacity: 0.3;
  cursor: default;
}

.c3 {
  margin: 1rem;
  min-width: 15rem;
  padding: 1.5rem 2.5rem;
  border-radius: 0.4rem;
  font-size: 1.6rem;
  font-weight: bold;
  border: yellow;
  background-color: #FF6347;
  color: #fff;
}

.c3:hover {
  background-color: #fff;
  color: #FF6347;
}

.c0 {
  max-width: 70rem;
  width: 100%;
  margin-top: 2rem 0;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: stretch;
  -webkit-box-align: stretch;
  -ms-flex-align: stretch;
  align-items: stretch;
  margin-bottom: 4rem;
}

@media (min-width:450px) {
  .c1 {
    -webkit-flex-direction: row;
    -ms-flex-direction: row;
    flex-direction: row;
    -webkit-box-pack: center;
    -webkit-justify-content: center;
    -ms-flex-pack: center;
    justify-content: center;
  }
}

<div>
  <div
    class="c0"
  >
    <div
      class="c1"
    >
      <button
        class="c2"
        disabled=""
      >
        Previous
      </button>
      <button
        class="c3"
      >
        Next
      </button>
    </div>
    <div
      data-testid="step"
    >
      Step 1
    </div>
  </div>
</div>
`
